### ISSUE TITLE
Native Crash: `_NativeCanvas.restore` throws EXC_BAD_ACCESS (KERN_INVALID_ADDRESS)

### ISSUE THREAD

**Issue by Sese-Schneider** [flutter/flutter#138607]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

According to Crashlytics, around 1% of our userbase is affected by the problem.
We're at the moment not sure about exact methods to reproduce.

1. Use an Apple device (iOS / iPadOS)

Noteable change: Occured first after upgrading from flutter `3.7.7` to `3.10.3` which brings Impeller to iOS by default.
Disabling impeller didn't stop the issue from appearing.

### Expected results

App doesn't crash.

### Actual results

App crashes.

### Code sample

https://github.com/zathras/jovial_svg/issues/78#issuecomment-1822597728

### Logs

<details><summary>Logs</summary>

> Crashed: io.flutter.1.ui
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x000000011daf7fc4

```console
Crashed: io.flutter.1.ui
0  Flutter                        0x33b974 (Missing UUID 4c4c441f55553144a1682f45551c1451)
1  App                            0xc9668 FfiTrampoline___pop$Method$FfiNative$Ptr + 1 (dart:ffi:1)
2  App                            0x9e30cc _NativeCanvas.restore + 5801 (painting.dart:5801)
3  App                            0xa84834 SIGroupHelper.endPaintGroup + 454 (common.dart:454)
4  App                            0xa84700 SIGroup.paint + 276 (array.dart:276)
5  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
6  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
7  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
8  App                            0x5c7a68 ScalableImageBase.paint + 812 (exported.dart:812)
9  App                            0x5c7854 _SIPainter.paint + 285 (widget.dart:285)
10 App                            0x259540 RenderCustomPaint._paintWithPainter + 609 (custom_paint.dart:609)
11 App                            0x259234 RenderCustomPaint.paint + 616 (custom_paint.dart:616)
12 App                            0xc226c RenderObject._paintWithContext + 3159 (object.dart:3159)
13 App                            0xa4efa8 PaintingContext.paintChild + 236 (object.dart:236)
14 App                            0x262498 RenderProxyBoxMixin.paint + 123 (proxy_box.dart:123)
15 App                            0xc226c RenderObject._paintWithContext + 3159 (object.dart:3159)
16 App                            0xc21b4 PaintingContext._repaintCompositedChild + 173 (object.dart:173)
17 App                            0xc1dc4 PipelineOwner.flushPaint + 192 (object.dart:192)
18 App                            0xb8728 RendererBinding.drawFrame + 498 (binding.dart:498)
19 App                            0xb7fb8 WidgetsBinding.drawFrame + 919 (binding.dart:919)
20 App                            0xb7984 RendererBinding._handlePersistentFrameCallback + 361 (binding.dart:361)
21 App                            0xb794c RendererBinding._handlePersistentFrameCallback + 359 (binding.dart:359)
22 App                            0x74738 SchedulerBinding._invokeFrameCallback + 558 (growable_array.dart:558)
23 App                            0x74668 SchedulerBinding.handleDrawFrame + 338 (iterable.dart:338)
24 App                            0x74380 SchedulerBinding._handleDrawFrame + 1067 (binding.dart:1067)
25 App                            0x74258 SchedulerBinding._handleDrawFrame + 1067 (binding.dart:1067)
26 App                            0x2a758 _rootRun + 1394 (zone.dart:1394)
27 App                            0x2a84c _rootRun + 1390 (zone.dart:1390)
28 App                            0x9e3db0 _CustomZone.run + 1297 (zone.dart:1297)
29 App                            0x9e4a88 _CustomZone.runGuarded + 1211 (zone.dart:1211)
30 App                            0x50150 _invoke + 165 (hooks.dart:165)
31 App                            0x50098 PlatformDispatcher._drawFrame + 400 (platform_dispatcher.dart:400)
32 App                            0x50054 _drawFrame + 138 (hooks.dart:138)
33 App                            0x50180 _drawFrame + 138 (hooks.dart:138)
34 App                            0xbc94 stub InvokeDartCode + 31124
35 Flutter                        0x61b920 (Missing UUID 4c4c441f55553144a1682f45551c1451)
36 Flutter                        0x7296d0 (Missing UUID 4c4c441f55553144a1682f45551c1451)
37 Flutter                        0x3e75a0 (Missing UUID 4c4c441f55553144a1682f45551c1451)
38 Flutter                        0x3c9990 (Missing UUID 4c4c441f55553144a1682f45551c1451)
39 Flutter                        0x3f97bc (Missing UUID 4c4c441f55553144a1682f45551c1451)
40 Flutter                        0x2de710 (Missing UUID 4c4c441f55553144a1682f45551c1451)
41 Flutter                        0x2e1d00 (Missing UUID 4c4c441f55553144a1682f45551c1451)
42 CoreFoundation                 0xcb624 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 32
43 CoreFoundation                 0x88f4c __CFRunLoopDoTimer + 940
44 CoreFoundation                 0x2e71c __CFRunLoopDoTimers + 288
45 CoreFoundation                 0x7a4d4 __CFRunLoopRun + 1852
46 CoreFoundation                 0x7f3ec CFRunLoopRunSpecific + 612
47 Flutter                        0x2e1dec (Missing UUID 4c4c441f55553144a1682f45551c1451)
48 Flutter                        0x2e1460 (Missing UUID 4c4c441f55553144a1682f45551c1451)
49 libsystem_pthread.dylib        0x16b8 _pthread_start + 148
50 libsystem_pthread.dylib        0xb88 thread_start + 8
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.13.9, on Microsoft Windows [Version 10.0.22631.2506], locale de-DE)
    • Flutter version 3.13.9 on channel stable at X:\Flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d211f42860 (3 weeks ago), 2023-10-25 13:42:25 -0700
    • Engine revision 0545f8705d
    • Dart version 3.1.5
    • DevTools version 2.25.0

[✓] Windows Version (Installed version of Windows is version 10 or higher)

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at X:\SDK
    • Platform android-34, build-tools 33.0.0
    • Java binary at: X:\Java\jdk-11.0.17\bin\java
    • Java version Java(TM) SE Runtime Environment 18.9 (build 11.0.17+10-LTS-269)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop Windows apps (Visual Studio Build Tools 2022 17.7.6)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools
    • Visual Studio Build Tools 2022 version 17.7.34221.43
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22631.2506]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 119.0.6045.124
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 119.0.2151.58

[✓] Network resources
    • All expected network resources are available.
```

</details>

**Comment by huycozy** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/138607#issue-1998751707 force flutter#v3.10.3 ios release arm64

**Comment by huycozy** [comment]
@flutter-symbolizer-bot this force ios flutter#v3.10.3 release arm64

```console
Crashed: io.flutter.1.ui
0  Flutter                        0x33b974 (Missing UUID 4c4c441f55553144a1682f45551c1451)
1  App                            0xc9668 FfiTrampoline___pop$Method$FfiNative$Ptr + 1 (dart:ffi:1)
2  App                            0x9e30cc _NativeCanvas.restore + 5801 (painting.dart:5801)
3  App                            0xa84834 SIGroupHelper.endPaintGroup + 454 (common.dart:454)
4  App                            0xa84700 SIGroup.paint + 276 (array.dart:276)
5  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
6  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
7  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
8  App                            0x5c7a68 ScalableImageBase.paint + 812 (exported.dart:812)
9  App                            0x5c7854 _SIPainter.paint + 285 (widget.dart:285)
10 App                            0x259540 RenderCustomPaint._paintWithPainter + 609 (custom_paint.dart:609)
11 App                            0x259234 RenderCustomPaint.paint + 616 (custom_paint.dart:616)
12 App                            0xc226c RenderObject._paintWithContext + 3159 (object.dart:3159)
13 App                            0xa4efa8 PaintingContext.paintChild + 236 (object.dart:236)
14 App                            0x262498 RenderProxyBoxMixin.paint + 123 (proxy_box.dart:123)
15 App                            0xc226c RenderObject._paintWithContext + 3159 (object.dart:3159)
16 App                            0xc21b4 PaintingContext._repaintCompositedChild + 173 (object.dart:173)
17 App                            0xc1dc4 PipelineOwner.flushPaint + 192 (object.dart:192)
18 App                            0xb8728 RendererBinding.drawFrame + 498 (binding.dart:498)
19 App                            0xb7fb8 WidgetsBinding.drawFrame + 919 (binding.dart:919)
20 App                            0xb7984 RendererBinding._handlePersistentFrameCallback + 361 (binding.dart:361)
21 App                            0xb794c RendererBinding._handlePersistentFrameCallback + 359 (binding.dart:359)
22 App                            0x74738 SchedulerBinding._invokeFrameCallback + 558 (growable_array.dart:558)
23 App                            0x74668 SchedulerBinding.handleDrawFrame + 338 (iterable.dart:338)
24 App                            0x74380 SchedulerBinding._handleDrawFrame + 1067 (binding.dart:1067)
25 App                            0x74258 SchedulerBinding._handleDrawFrame + 1067 (binding.dart:1067)
26 App                            0x2a758 _rootRun + 1394 (zone.dart:1394)
27 App                            0x2a84c _rootRun + 1390 (zone.dart:1390)
28 App                            0x9e3db0 _CustomZone.run + 1297 (zone.dart:1297)
29 App                            0x9e4a88 _CustomZone.runGuarded + 1211 (zone.dart:1211)
30 App                            0x50150 _invoke + 165 (hooks.dart:165)
31 App                            0x50098 PlatformDispatcher._drawFrame + 400 (platform_dispatcher.dart:400)
32 App                            0x50054 _drawFrame + 138 (hooks.dart:138)
33 App                            0x50180 _drawFrame + 138 (hooks.dart:138)
34 App                            0xbc94 stub InvokeDartCode + 31124
35 Flutter                        0x61b920 (Missing UUID 4c4c441f55553144a1682f45551c1451)
36 Flutter                        0x7296d0 (Missing UUID 4c4c441f55553144a1682f45551c1451)
37 Flutter                        0x3e75a0 (Missing UUID 4c4c441f55553144a1682f45551c1451)
38 Flutter                        0x3c9990 (Missing UUID 4c4c441f55553144a1682f45551c1451)
39 Flutter                        0x3f97bc (Missing UUID 4c4c441f55553144a1682f45551c1451)
40 Flutter                        0x2de710 (Missing UUID 4c4c441f55553144a1682f45551c1451)
41 Flutter                        0x2e1d00 (Missing UUID 4c4c441f55553144a1682f45551c1451)
42 CoreFoundation                 0xcb624 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 32
43 CoreFoundation                 0x88f4c __CFRunLoopDoTimer + 940
44 CoreFoundation                 0x2e71c __CFRunLoopDoTimers + 288
45 CoreFoundation                 0x7a4d4 __CFRunLoopRun + 1852
46 CoreFoundation                 0x7f3ec CFRunLoopRunSpecific + 612
47 Flutter                        0x2e1dec (Missing UUID 4c4c441f55553144a1682f45551c1451)
48 Flutter                        0x2e1460 (Missing UUID 4c4c441f55553144a1682f45551c1451)
49 libsystem_pthread.dylib        0x16b8 _pthread_start + 148
50 libsystem_pthread.dylib        0xb88 thread_start + 8
```

**Comment by huycozy** [comment]
@Sese-Schneider Could you confirm you built the app with Flutter 3.10.3? (as I see your flutter doctor is on 3.13.9).

Also, please provide the list of affected device info as well (OS, model). Thanks!

**Comment by Sese-Schneider** [comment]
@huycozy the issue happens on both, Flutter `3.10.3` and `3.13.9`
Example devices are:
- iPad 8th gen (iPadOS 16.3.1)
- iPhone 13 Mini (iOS 16.7.2)
- iPhone 12 (iOS 17.1.1)
- ... A lot more devices and OS versions
There seems to be no clear pattern in devices affected, instead it beeing a general issue with apple OSes

**Comment by huycozy** [comment]
@flutter-symbolizer-bot this force ios flutter#v3.13.9 release arm64

```
Crashed: io.flutter.1.ui
0  Flutter                        0x33b974 (Missing UUID 4c4c441f55553144a1682f45551c1451)
1  App                            0xc9668 FfiTrampoline___pop$Method$FfiNative$Ptr + 1 (dart:ffi:1)
2  App                            0x9e30cc _NativeCanvas.restore + 5801 (painting.dart:5801)
3  App                            0xa84834 SIGroupHelper.endPaintGroup + 454 (common.dart:454)
4  App                            0xa84700 SIGroup.paint + 276 (array.dart:276)
5  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
6  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
7  App                            0xa84784 SIGroup.paint + 430 (dag.dart:430)
8  App                            0x5c7a68 ScalableImageBase.paint + 812 (exported.dart:812)
9  App                            0x5c7854 _SIPainter.paint + 285 (widget.dart:285)
10 App                            0x259540 RenderCustomPaint._paintWithPainter + 609 (custom_paint.dart:609)
11 App                            0x259234 RenderCustomPaint.paint + 616 (custom_paint.dart:616)
12 App                            0xc226c RenderObject._paintWithContext + 3159 (object.dart:3159)
13 App                            0xa4efa8 PaintingContext.paintChild + 236 (object.dart:236)
14 App                            0x262498 RenderProxyBoxMixin.paint + 123 (proxy_box.dart:123)
15 App                            0xc226c RenderObject._paintWithContext + 3159 (object.dart:3159)
16 App                            0xc21b4 PaintingContext._repaintCompositedChild + 173 (object.dart:173)
17 App                            0xc1dc4 PipelineOwner.flushPaint + 192 (object.dart:192)
18 App                            0xb8728 RendererBinding.drawFrame + 498 (binding.dart:498)
19 App                            0xb7fb8 WidgetsBinding.drawFrame + 919 (binding.dart:919)
20 App                            0xb7984 RendererBinding._handlePersistentFrameCallback + 361 (binding.dart:361)
21 App                            0xb794c RendererBinding._handlePersistentFrameCallback + 359 (binding.dart:359)
22 App                            0x74738 SchedulerBinding._invokeFrameCallback + 558 (growable_array.dart:558)
23 App                            0x74668 SchedulerBinding.handleDrawFrame + 338 (iterable.dart:338)
24 App                            0x74380 SchedulerBinding._handleDrawFrame + 1067 (binding.dart:1067)
25 App                            0x74258 SchedulerBinding._handleDrawFrame + 1067 (binding.dart:1067)
26 App                            0x2a758 _rootRun + 1394 (zone.dart:1394)
27 App                            0x2a84c _rootRun + 1390 (zone.dart:1390)
28 App                            0x9e3db0 _CustomZone.run + 1297 (zone.dart:1297)
29 App                            0x9e4a88 _CustomZone.runGuarded + 1211 (zone.dart:1211)
30 App                            0x50150 _invoke + 165 (hooks.dart:165)
31 App                            0x50098 PlatformDispatcher._drawFrame + 400 (platform_dispatcher.dart:400)
32 App                            0x50054 _drawFrame + 138 (hooks.dart:138)
33 App                            0x50180 _drawFrame + 138 (hooks.dart:138)
34 App                            0xbc94 stub InvokeDartCode + 31124
35 Flutter                        0x61b920 (Missing UUID 4c4c441f55553144a1682f45551c1451)
36 Flutter                        0x7296d0 (Missing UUID 4c4c441f55553144a1682f45551c1451)
37 Flutter                        0x3e75a0 (Missing UUID 4c4c441f55553144a1682f45551c1451)
38 Flutter                        0x3c9990 (Missing UUID 4c4c441f55553144a1682f45551c1451)
39 Flutter                        0x3f97bc (Missing UUID 4c4c441f55553144a1682f45551c1451)
40 Flutter                        0x2de710 (Missing UUID 4c4c441f55553144a1682f45551c1451)
41 Flutter                        0x2e1d00 (Missing UUID 4c4c441f55553144a1682f45551c1451)
42 CoreFoundation                 0xcb624 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 32
43 CoreFoundation                 0x88f4c __CFRunLoopDoTimer + 940
44 CoreFoundation                 0x2e71c __CFRunLoopDoTimers + 288
45 CoreFoundation                 0x7a4d4 __CFRunLoopRun + 1852
46 CoreFoundation                 0x7f3ec CFRunLoopRunSpecific + 612
47 Flutter                        0x2e1dec (Missing UUID 4c4c441f55553144a1682f45551c1451)
48 Flutter                        0x2e1460 (Missing UUID 4c4c441f55553144a1682f45551c1451)
49 libsystem_pthread.dylib        0x16b8 _pthread_start + 148
50 libsystem_pthread.dylib        0xb88 thread_start + 8
```

**Comment by huycozy** [comment]
The bot can't seem to symbolize the stack trace. I will label the issue for further insights.

**Comment by Sese-Schneider** [comment]
@huycozy we've managed to reproduce the issue and have a minimal reproducable sample.

the crash is caused by a third-party dependency, but I still think that the underlying issue is with flutter and not an implementation error they've made.

See how to reproduce here: https://github.com/zathras/jovial_svg/issues/78

**Comment by Sese-Schneider** [comment]
The maintainers of `jovial_svg` do confirm that it's not a bug with their package, but with the native code that needs to be fixed.
`jovial_svg` properly balances save/restore calls on the asset: https://github.com/zathras/jovial_svg/issues/78#issuecomment-1823823796

@huycozy

**Comment by huycozy** [comment]
Hi @Sese-Schneider 
Could you share a complete sample code? I see there is one [here](https://github.com/zathras/jovial_svg/issues/78#issuecomment-1822597728) but there is only one svg asset whilst the code needs an array of assets `assets/test$assetIndex.svg`.

**Comment by Sese-Schneider** [comment]
### Sample code

In your `pubspec.yaml` add the dependency `jovial_svg`.

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:jovial_svg/jovial_svg.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'SVG rendering crash Demo',
      home: Scaffold(
        body: CustomScrollView(
          slivers: [
            const SliverAppBar(
              title: Text('SVG rendering crash Demo'),
            ),
            SliverList.builder(
              itemBuilder: (context, index) => SizedBox(
                key: ValueKey(index),
                height: 300,
                child: (index % 4 == 0)
                    ? _buildSVGWidget()
                    : ColoredBox(
                        color: Color.fromARGB(
                            0xff, 64 * ((index % 4) + 1), 0x88, 0xaa)),
              ),
            ),
          ],
        ),
      ),
    );
  }

  ScalableImageWidget _buildSVGWidget() {
    final svgSource =
        ScalableImageSource.fromSvg(rootBundle, 'assets/test1.svg');
    return ScalableImageWidget.fromSISource(
      si: svgSource,
      fit: BoxFit.contain,
      isComplex: true,
    );
  }
}
```

Asset `test1.svg`:

 ![test1.svg](https://github.com/zathras/jovial_svg/assets/14337484/6c697eb7-07d8-4a0a-aa4a-043bfcbed328)
 
 cc: @huycozy

**Comment by huycozy** [comment]
@Sese-Schneider Thanks for the sample code, I checked it on Iphone 15 Pro, iOS 17.0 (emulator) but couldn't reproduce the crash. The list is very laggy while scrolling, though.

<details>
<summary>Demo video</summary>

https://github.com/flutter/flutter/assets/104349824/13b692f0-5f28-4c3c-a0e8-77a760411651
</details>

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.16.1, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.16.1 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7f20e5d18c (9 hours ago), 2023-11-27 09:47:30 -0800
    • Engine revision 22b600f240
    • Dart version 3.2.1
    • DevTools version 2.28.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version unknown)
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.8+0-17.0.8b1000.22-10799086)

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio Giraffe Patch 3.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.84.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 119.0.6045.159

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

```bash
[!] Flutter (Channel master, 3.17.0-16.0.pre.34, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.17.0-16.0.pre.34 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2150424cee (20 minutes ago), 2023-11-27 18:23:21 -0800
    • Engine revision 96137d05fa
    • Dart version 3.3.0 (build 3.3.0-169.0.dev)
    • DevTools version 2.30.0-dev.4
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version unknown)
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.8+0-17.0.8b1000.22-10799086)

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio Giraffe Patch 3.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.84.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 119.0.6045.159

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```
</details>

**Comment by stuartmorgan** [comment]
Re-routing to the engine team; the end of this stack is in the engine's canvas wrapper code.

**Comment by flar** [comment]
Using the test case above I was able to track this down to the same underlying root cause as we were seeing inhttps://github.com/flutter/flutter/issues/149340 and the problem will be fixed in the next stable release. See https://github.com/flutter/flutter/blob/master/docs/releases/Flutter-build-release-channels.md#stable

**Comment by flar** [comment]
Closing as per comments by @jmagman on https://github.com/flutter/flutter/issues/149340

Please verify that you no longer see this on the beta channel and reopen if it still reproduces there.

**Comment by isegal** [comment]
@flar is it possible to cherry-pick the fix? Which commit is the fix for the issue?

**Comment by flar** [comment]
> @flar is it possible to cherry-pick the fix? Which commit is the fix for the issue?

The bug is no longer reproducible in the latest code and that code is now constructed in such a way as to prevent accidental stale pointer usage. However, these changes were not written to fix this bug, they were part of a much broader redesign to make it easier for future work to optimize the DisplayList. There are 2 problems in trying to identify something that could be considered a "cherry-pick" fix:

- The redesign took place over a couple of very large commits, even if we bisected to the exact commit that made the crash no longer reproducible, the code that would be the target of the cherry-pick would be too risky and would intersect with too many other fixes along the way, increasing the number of commits and files that would have to be charry-picked.
- At least one of those "redesign" commits had to be reverted and re-landed at least once, meaning that this would be a change that is risky for both its size and the fact that it wasn't simple enough to get right in one PR.

**Comment by isegal** [comment]
Thank you for the detailed response. It seems that puts us into a bad situation - if we roll out a 3.22-based app we risk an outage. If we consume the beta channel, that is risky too.

I would like to propose for Flutter to consider having an LTS version that does not move at a risky pace of changes.

**Comment by stuartmorgan** [comment]
> I would like to propose for Flutter organization to consider having an LTS version that does not move at a risky pace of changes.

Anyone is welcome to [file new feature requests](https://github.com/flutter/flutter/issues/new/choose).

**Comment by flar** [comment]
The issue seems to have been fixed in the latest stable (3.24) as per my comment on a related bug. Please try our apps again on stable (3.24) and see if the problem is fixed for you.

https://github.com/flutter/flutter/issues/149305#issuecomment-2285327305
